draft-ietf-tls-tls13.md

```diff
@@ -965,14 +968,10 @@ type
 : The higher-level protocol used to process the enclosed fragment.
 
 version
-: The version of the protocol being employed.  This document
-  describes TLS Version 1.3, which uses the version { 3, 4 }.  The
-  version value 3.4 is historical, deriving from the use of {3, 1}
-  for TLS 1.0.  (See {{record-layer-1}}.)  Note that a client that
-  supports multiple versions of TLS may not know what version will
-  be employed before it receives the ServerHello.  See
-  {{backward-compatibility}} for discussion about what record layer
-  version number should be employed for ClientHello.
+: The version of the protocol being employed in the current record.
+  TLS clients MUST set this version to { 3, 1 } for the initial ClientHello.
+  For the ServerHello and all other TLS records this MUST be equal to the negotiated version.
```

Comment:
This rule isn't sufficient to hammer down the record-layer version. There's also alerts sent:
* By the server in response to a protocol error in trying to read ClientHello.
* By the client in response to a protocol error in trying to read ServerHello.
* By the server in response to ClientHello with a bad version
* By the client in response to ServerHello with a bad version
(And conversely what versions the peer will accept.)

The first two are unambiguously unnegotiated, so I think either we shouldn't bother sending anything and drop the connection or the version should be {3, 1}.

Making the third {3, 1} is nice and consistent with the other unnegotiated cases. The client should accept any version (it hasn't pinned a version yet anyway). Also most server implementations will resolve the version before the cipher suite, and it would be unnecessary complexity on the server's end to require that an alert on no valid cipher suite still use {3, 1} rather than the version it negotiated.

The fourth is tricky. If the client were to send protocol_version because the version is too low and we wish for the server to actually read it, at this point the server has already switched versions. So it won't accept any record with something other than the version the server negotiated. (And recommending new servers accept that is unnecessary complication in the state machine.) Thus it should be the version the server negotiated and client implementations should take care that, when programming their record layer to echo that version, that it be treated entirely opaque and never condition on it. (Lest it, say, accidentally act on being "in SSLv3" when SSLv3 is disabled or unimplemented.)

Hammering these down would avoid the silliness where FALLBACK_SCSV's spec has its own language to specify the record-layer version.